Add client dashboard

hihuz opened the pull request:
In this PR I am adding the main screen after login which will be a list of existing oAuth clients with some possible actions (create client, click on client to get details.. these screens will come in upcoming PRs)

I used https://www.figma.com/file/QsjOR6Uyi6CE20aLN9Fe2pvf/web-app_new?node-id=1689%3A4503 as inspiration for design:

<img width="1035" alt="Screenshot 2019-10-23 at 12 55 16" src="https://user-images.githubusercontent.com/3253522/67386243-de985180-f594-11e9-9fac-7792e11d5ad0.png">


if you want to test locally, `cd console-src && yarn && yarn start` and you will need a client in your clients table with id of `fe8055d9-f496-49e9-9551-5b416139630e` (with authorization_code clients_grant_types)

hihuz committed:
Add client dashboard

hihuz commented on console-src/src/enhancers/withOAuthClients.tsx:
```diff
@@ -0,0 +1,190 @@
+        (state: State): State => ({
+          ...state,
+          isLoading: false,
+          // Currently rawQuery return type does not include clients
```
I have opened a PR against the sdk to fix this: https://github.com/kontist/sdk/pull/7

hihuz commented on console-src/src/queries/index.tsx:
```diff
@@ -0,0 +1,19 @@
+  }
+`;
+
+export const createClientMutation = ``;
```
This is of course temporary, I will fill in these queries as I implement the corresponding UI elements

dmytroyarmak commented on console-src/src/components/LinkButton/LinkButton.tsx:
```diff
@@ -0,0 +1,31 @@
+  color: white;
+  width: 150px;
+  min-width: max-content;
+  height: 50px;
```
What would happen when text is wrapped to the next line? Do we need to prevent it and add ellipsis?

hihuz replied:
Maybe yeah, I have shamelessly copy pasted these styles from the webapp but I don't see why we can't improve on them 👍

hihuz replied:
Actually the `width` property is not respected when text goes above 150px, because `<a>` generated by `Link` is an `inline` element.

I think I will just remove the `width` here, and try to keep button texts short, I think it's important that whole button text are visible

hihuz replied:
I have done it here: 40cfca52cc3ee86ac7ce7776de0535682ae088df I think it's the best approach for this

dmytroyarmak replied:
awesome, thanks

dmytroyarmak commented on console-src/src/components/ClientList/ClientList.tsx:
```diff
@@ -0,0 +1,80 @@
+  border-top-left-radius: 0.25rem;
+  border-top-right-radius: 0.25rem;
+  padding: 1rem;
+  height: 30px;
```
Can text in `ClientListHeader` be wrapped to the next line?

hihuz replied:
I don't think it would happen currently as this is some static text, but I agree with you that it should be a bit more flexible, I'll tweak the styles to handle these overflowing text cases

hihuz replied:
I have changed to `min-height` in this commit 40cfca52cc3ee86ac7ce7776de0535682ae088df so that if text is overflowing it will wrap and the container will expand its height

dmytroyarmak commented on console-src/src/enhancers/withOAuthClients.tsx:
```diff
@@ -0,0 +1,190 @@
+  constructor(props: {}) {
+    super(props);
+
+    this.fetchClients = async () => {
```
Why are this methods defined in constructor instead of just having them in class like:
```
constructor() {
  ...
}

fetchClients = async () => {
...
}
```
?

hihuz replied:
this is the recommended way in React docs to use context when you want to pass methods to consumers: https://reactjs.org/docs/context.html#updating-context-from-a-nested-component

I'm not 100% sure that it is required to do it this way, and it looks a little odd, but I'd rather stick to the official recommendation, what do you think?

hihuz replied:
Actually on second thought I think I will try to use regular methods as you suggested and if it still works properly I'll stick to that, it would be less confusing for everybody

hihuz replied:
So I have tried and it seems to work with regular methods 🤷‍♂ so let's go with that for now, thanks for the suggestion 👍 

changed here: a7bf1ea8d0780811180cdae92c60404325cc521c

dmytroyarmak commented on console-src/src/enhancers/withOAuthClients.tsx:
```diff
@@ -0,0 +1,190 @@
+      const { viewer } = await client.graphQL.rawQuery(fetchClientsQuery);
+
+      this.setState(
+        (state: State): State => ({
```
Why is callback version of `setState` is used here?

hihuz replied:
Ah true it's not so useful now, initially I had some nested values which required me to use this callback form, I will update

hihuz replied:
I have changed it here: 950aca0f7897526eab39dfc3dd4cbbd31d4e5526

dmytroyarmak commented on console-src/src/enhancers/withOAuthClients.tsx:
```diff
@@ -0,0 +1,190 @@
+      isLoading: false,
+      createClient: this.createClient,
+      updateClient: this.updateClient,
+      deleteClient: this.deleteClient,
```
Why are these methods stored in component's state?

hihuz replied:
Same as above 🤷‍♂ 
https://reactjs.org/docs/context.html#updating-context-from-a-nested-component

It is also quite confusing to me but on the other hand I have always done it this way when using the most recent context api

hihuz replied:
same as above, it seems that it is not necessary, so I have changed it here: a7bf1ea8d0780811180cdae92c60404325cc521c

dmytroyarmak commented on console-src/src/enhancers/withOAuthClients.tsx:
```diff
@@ -0,0 +1,190 @@
+  deleteClient: () => {}
+});
+
+class OAuthClientsProvider extends Component<{}, State> {
```
It looks like this component has two responsibilities:
1. Authorize user and fetch access token
2. Provide context to handle oauth clients

Can these responsibilities be separated into 2 components?

hihuz replied:
Yeah it is certainly possible and it also looks more reasonable to me, I will change it 👍

hihuz replied:
I have in the end separated the logic in 3 different components:
1. `KontistClientProvider` will just create a `kontistClient` instance and expose it to the components that need it through render props (we only want to have one single instance of it for the whole app)
2. `AuthenticateUser` will have `kontistClient` injected as prop and perform authentication tasks
3. `OAuthClientsProvider` will have `kontistClient` injected as prop and perform client fetching and storing and forward them to consumers who need them

Does it look better to you? (it does to me 😅 thank you for the suggestion) 73319c8a2ded4e3b5e9c48e9da30e8c14fb0918a

dmytroyarmak replied:
Yeah, this separation is even better. Good job! 👍

dmytroyarmak commented on console-src/src/layout/Layout.tsx:
```diff
@@ -0,0 +1,44 @@
+
+const Content = styled.main`
+  width: 660px;
+`;
```
Should it be responsive and be suitable for devices with width less than 660px?

hihuz replied:
Yes, it should 🙂 I will update it

hihuz replied:
I have changed it here: 40cfca52cc3ee86ac7ce7776de0535682ae088df using `max-width`, it looks like this on small device:

<img width="351" alt="Screenshot 2019-10-23 at 16 25 42" src="https://user-images.githubusercontent.com/3253522/67403106-c8e55500-f5b1-11e9-9ae9-7398a71e4468.png">

hihuz committed:
Tweak styles to be more flexible

hihuz committed:
Do not use setState callback when not necessary

hihuz committed:
Extract some enhancers to separate concerns

hihuz committed:
Use regular class methods for context updaters

hihuz committed:
Remove unused components

They were backported from the webapp but are not used for now, we will reintroduce them as needed

dmytroyarmak reviewed (approved):
awesome 🔥

johannespfeiffer commented:
Very cool!
Minor thing: Can we adjust the wording to "OAuth2" to be more consistent and specific?
<img width="498" alt="Bildschirmfoto 2019-10-23 um 19 22 24" src="https://user-images.githubusercontent.com/6780316/67418273-c7745680-f5ca-11e9-985c-d0f0431d0d53.png">

hihuz committed:
Unify OAuth2 wording

hihuz commented:
Ah true thanks @johannespfeiffer I have changed it here: d5151f96f0c633c83222a7dccf06d3ecf658e65f
